pkg/cleaner/cleaner.go

```diff
@@ -2,131 +2,33 @@
-	c.mut.Unlock()
-}
+// Run starts the deleted app cleaner. Every pollSleepDuration, it compares the result of nsLister.List with the directories in the top level of gitHome on the local file system. On any error, it uses log messages to output a human readable description of what happened.
+func Run(gitHome string, nsLister kcl.NamespaceInterface) error {
 
-// localDirs returns all of the local directories immediately under gitHome that filter returns true for. filter will receive only the names of each of the top level directories (not their fully qualified paths), and should return true if it should be included in the output
-func localDirs(gitHome string, filter func(string) bool) ([]string, error) {
-	fileInfos, err := ioutil.ReadDir(gitHome)
+	watcher, err := nsLister.Watch(nil, nil, "")
 	if err != nil {
-		return nil, err
-	}
-	var ret []string
-	for _, fileInfo := range fileInfos {
-		nm := fileInfo.Name()
-		if len(nm) <= 0 || nm == "." || !fileInfo.IsDir() {
-			continue
-		}
-		if filter(nm) {
-			ret = append(ret, filepath.Join(gitHome, nm))
-		}
-	}
-	return ret, nil
-}
-
-// getDiff gets the directories that are not in namespaceList
-func getDiff(namespaceList []api.Namespace, dirs []string) []string {
-	var ret []string
-
-	// create a set of lowercase namespace names
-	namespacesSet := make(map[string]struct{})
-	for _, ns := range namespaceList {
-		lowerName := strings.ToLower(ns.Name)
-		namespacesSet[lowerName] = struct{}{}
-	}
-
-	// get dirs not in the namespaces set
-	for _, dir := range dirs {
-		lowerName := strings.ToLower(dir)
-		if _, ok := namespacesSet[lowerName]; !ok {
-			ret = append(ret, lowerName)
-		}
+		log.Printf("unable to get watch events (%s)", err)
 	}
-
-	return ret
-}
-
-func stripSuffixes(strs []string, suffix string) []string {
-	ret := make([]string, len(strs))
-	for i, str := range strs {
-		idx := strings.LastIndex(str, suffix)
-		if idx >= 0 {
-			ret[i] = str[:idx]
-		} else {
-			ret[i] = str
-		}
-	}
-	return ret
-}
-
-func dirHasGitSuffix(dir string) bool {
-	return strings.HasSuffix(dir, dotGitSuffix)
-}
-
-// Run starts the deleted app cleaner. Every pollSleepDuration, it compares the result of nsLister.List with the directories in the top level of gitHome on the local file system. On any error, it uses log messages to output a human readable description of what happened.
-func (c Ref) Run(gitHome string, nsLister k8s.NamespaceLister, ref Ref, pollSleepDuration time.Duration) error {
 	for {
-		nsList, err := nsLister.List(labels.Everything(), fields.Everything())
-		if err != nil {
-			log.Printf("Cleaner error listing namespaces (%s)", err)
-			continue
-		} else {
-			lst := make([]string, len(nsList.Items))
-			for i, ns := range nsList.Items {
-				lst[i] = strings.ToLower(ns.Name)
+		event := <-watcher.ResultChan()
+		namespace := event.Object.(*api.Namespace)
+		if event.Type == "DELETED" {
+			if err := os.RemoveAll(namespace.ObjectMeta.Name + dotGitSuffix); err != nil {
```

Comment:
If a folder is deleted while a `git push` operation is in progress, the behavior of the builder is essentially undefined and the user will see one of quite a few error messages, none of which make sense. I saw this behavior in effect when I first built the cleaner, before there was locking.

That needs to be fixed in some way. The most obvious way is to use a lock, but since the description here says `removes locking completely for cleaner`, I'm guessing you don't want to go there ;). Instead, perhaps you could use a `WaitGroup` that the cleaner increments when it's about to do its work, decrements when it's done, and the SSH server waits for?